File: frontend/src/api/stats.ts
import api from './client';

/** Basic app stats (totals and top 5 lists) */
export interface AppStats {
  total_entries: number;
  total_decks: number;
  total_cards: number;
  total_spreads: number;
  top_decks: Array<[string, number]>;
  top_spreads: Array<[string, number]>;
}

/** Extended stats with additional metrics for the Stats tab */
export interface ExtendedStats extends AppStats {
  entries_this_month: number;
  total_readings: number;
  unique_cards_drawn: number;
  avg_cards_per_reading: number;
}

/** Card frequency data for visualization */
export interface CardFrequency {
  name: string;
  count: number;
  reversed_count: number;
}

/** Timeline data point (one per month) */
export interface TimelinePeriod {
  period: string;
  entries: number;
  readings: number;
}

/** Tag usage count */
export interface TagTrend {
  name: string;
  color: string;
  count: number;
}

/** Single usage item (deck or spread) */
export interface UsageItem {
  name: string;
  count: number;
}

/** Deck and spread usage stats */
export interface UsageStats {
  top_decks: UsageItem[];
  top_spreads: UsageItem[];
}

export async function getStats(): Promise<AppStats> {
  const res = await api.get('/api/stats');
  return res.data;
}

export async function getExtendedStats(): Promise<ExtendedStats> {
  const res = await api.get('/api/stats/extended');
  return res.data;
}

export async function getCardFrequency(
  limit?: number,
  deckId?: number
): Promise<CardFrequency[]> {
  const params = new URLSearchParams();
  if (limit) params.set('limit', String(limit));
  if (deckId) params.set('deck_id', String(deckId));

  const query = params.toString();
  const res = await api.get(`/api/stats/card-frequency${query ? `?${query}` : ''}`);
  return res.data;
}

export async function getTimeline(limit?: number): Promise<TimelinePeriod[]> {
  const query = limit ? `?limit=${limit}` : '';
  const res = await api.get(`/api/stats/timeline${query}`);
  return res.data;
}

export async function getTagTrends(limit?: number): Promise<TagTrend[]> {
  const query = limit ? `?limit=${limit}` : '';
  const res = await api.get(`/api/stats/tag-trends${query}`);
  return res.data;
}

export async function getUsageStats(limit?: number): Promise<UsageStats> {
  const query = limit ? `?limit=${limit}` : '';
  const res = await api.get(`/api/stats/usage${query}`);
  return res.data;
}

File: backend/routes/stats.py
"""
Statistics endpoints for the Stats tab.
"""

from flask import Blueprint, jsonify, current_app, request

stats_bp = Blueprint('stats', __name__)


@stats_bp.route('/api/stats')
def get_stats():
    """Get basic statistics (totals and top 5 lists)."""
    db = current_app.config['DB']
    data = db.get_stats()
    return jsonify(data)


@stats_bp.route('/api/stats/extended')
def get_extended_stats():
    """Get extended statistics for the Stats tab overview.

    Returns basic stats plus entries_this_month, unique_cards_drawn,
    total_readings, and avg_cards_per_reading.
    """
    db = current_app.config['DB']
    data = db.get_extended_stats()
    return jsonify(data)


@stats_bp.route('/api/stats/card-frequency')
def get_card_frequency():
    """Get card frequency data for visualization.

    Query params:
        limit: Max cards to return (default 20, max 100)
        deck_id: Optional deck ID to filter by

    Returns list of {name, deck_name, count, reversed_count}
    """
    db = current_app.config['DB']

    # Parse query params with validation
    limit = request.args.get('limit', 20, type=int)
    limit = min(max(limit, 1), 100)  # Clamp between 1 and 100

    deck_id = request.args.get('deck_id', type=int)

    data = db.get_card_frequency(limit=limit, deck_id=deck_id)
    return jsonify(data)


@stats_bp.route('/api/stats/timeline')
def get_timeline():
    """Get entry and reading counts grouped by month.

    Query params:
        limit: Number of months to return (default 12, max 36)
    """
    db = current_app.config['DB']
    limit = request.args.get('limit', 12, type=int)
    limit = min(max(limit, 1), 36)
    data = db.get_timeline_stats(limit=limit)
    return jsonify(data)


@stats_bp.route('/api/stats/tag-trends')
def get_tag_trends():
    """Get entry tag usage counts.

    Query params:
        limit: Max tags to return (default 15, max 50)
    """
    db = current_app.config['DB']
    limit = request.args.get('limit', 15, type=int)
    limit = min(max(limit, 1), 50)
    data = db.get_tag_trends(limit=limit)
    return jsonify(data)


@stats_bp.route('/api/stats/usage')
def get_usage_stats():
    """Get deck and spread usage counts.

    Query params:
        limit: Max items per category (default 10, max 50)
    """
    db = current_app.config['DB']
    limit = request.args.get('limit', 10, type=int)
    limit = min(max(limit, 1), 50)
    data = db.get_usage_stats(limit=limit)
    return jsonify(data)
